mautic/mautic
# Reports: smarter query generation for faster load times

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | Speed enhancement
| Automated tests included? | Y
| Related user documentation PR URL | /
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

~~⚠️ This PR modifies a file added in https://github.com/mautic/mautic/pull/7088. Merge that PR first!~~ - DONE


Reports join some tables even though the report configuration doesn't require to select, group by, order by nor filter by any column in that joined table. This PR makes the reports smarter so they would join the tables only if necessary. This will speed up some reports significantly.

This report was very slow for one of our customers:
![Screenshot 2019-03-12 at 16 11 22](https://user-images.githubusercontent.com/1235442/54211537-cda53f00-44e1-11e9-9e25-63ee739b4eaa.png)

It generated SQL query like this:
```
SELECT es.email_address             AS email_address, 
       IF(dnc.id IS NOT NULL 
          AND dnc.reason = 2, 1, 0) AS bounced, 
       Date(es.date_read)           AS date_read 
FROM   mautic_email_stats es 
       LEFT JOIN mautic_emails e 
              ON e.id = es.email_id 
       LEFT JOIN mautic_leads l 
              ON l.id = es.lead_id 
       LEFT JOIN mautic_ip_addresses i 
              ON i.id = es.ip_id 
       LEFT JOIN mautic_emails vp 
              ON vp.id = e.variant_parent_id 
       LEFT JOIN mautic_categories c 
              ON c.id = e.category_id 
       LEFT JOIN mautic_lead_donotcontact dnc 
              ON e.id = dnc.channel_id 
                 AND dnc.channel = 'email' 
                 AND es.lead_id = dnc.lead_id 
WHERE  ( es.date_sent IS NULL 
          OR ( es.date_sent BETWEEN '2016-12-16 23:00:00' AND '2019-01-16 22:59:59' ) ) 
       AND ( es.date_read >= '2018-02-25 15:20:00' ) 
ORDER  BY es.date_read ASC 
LIMIT  30 offset 0 
```

With this PR checked out the query looks like this:
```
SELECT es.email_address             AS email_address, 
       IF(dnc.id IS NOT NULL 
          AND dnc.reason = 2, 1, 0) AS bounced, 
       Date(es.date_read)           AS date_read 
FROM   mautic_email_stats es 
       LEFT JOIN mautic_lead_donotcontact dnc 
              ON es.email_id = dnc.channel_id 
                 AND dnc.channel = 'email' 
                 AND es.lead_id = dnc.lead_id 
WHERE  ( es.date_sent IS NULL 
          OR ( es.date_sent BETWEEN '2016-12-16 23:00:00' AND '2019-03-13 22:59:59' ) ) 
       AND ( es.date_read >= '2019-03-11 08:45:46' ) 
ORDER  BY es.date_read ASC 
LIMIT  30 offset 0 
```

In this example we can remove 5 out of 6 joins and the result will be the same but faster.

#### Steps to test this PR:
This PR needs thorough testing as it may break for any report source with any configuration. The bigger the database the better. Also check that the displayed data in the reports make sense.

1. Create report for each report source and test random fields in select, order, group and filter.
